ENIGMAsystem/SHELL/Universal_System/Extensions/Asynchronous/ASYNCdialog.cpp

```diff
@@ -118,9 +118,10 @@ static void* getLoginAsync(void* data) {
   threads[md->id]->active = false;
   ds_map_replaceanyway(async_load, "id", md->id);
   string ret = threads[md->id]->ret;
-  size_t end = ret.find('\0', 0);
-  ds_map_replaceanyway(async_load, "username", ret.substr(0, end));
-  ds_map_replaceanyway(async_load, "password", ret.substr(end + 1, ret.size() - end));
+  vector<string> split = string_split(ret, '\0');
+  // must still check if the size is larger than 0 for when user cancels the dialog
+  ds_map_replaceanyway(async_load, "username", (split.size() > 0) ? split[0] : "");
+  ds_map_replaceanyway(async_load, "password", (split.size() > 1) ? split[1] : "");
```

Comment:
I think you'll find I did start a review. You'll get used to the new review system with time; it's a little janky. As I said, it seems to be based on review tools such as Critique at Google, but it's missing a few features that make Critique so usable. I expect them to fill in over time.

I left a comment on your `string_split` solution—the vector will "contain" only one element if there's no null in the string, so your code only survives because there's no error reporting mechanism. If you enabled STL's debug mode, you'd find it suddenly explodes. And rightfully so—you're directly accessing `vec[0]` and `vec[1]`. From now on, any time you see a literal array access like those, I want you to panic inside until you verify that they're guarded by something like `if (vec.size() > 1)`. Which yours aren't.

It's more natural (and more efficient) to protect your `find()`-based solution, so I recommend doing that. As for users, it might be confusing to them at first, I agree. I wanted to solve this with a proper array type in ENIGMA. It could be the return type of this method, and `var` could have a proper constructor for it. Alternatively, we can drop GM6's behavior of `var = var` performing a shallow copy. If not for that design decision, `var` would be a perfectly fine mechanism for returning tuples.

I think the GM:S behavior of replace is better, and am happy to migrate to it. There's probably some application that depends on the old behavior, but it's not a very defensible behavior.

Reply:
Josh, mine does already work that way, look again: ``(split.size() > 1) ? split[1] : ""``

Anyway, I'm going to change it just because the other split method is less verbose and more straight forward.

I have no opinion on preserving the GM6 behavior because I don't know how widespread that behavior was relied upon. I doubt that it was relied upon very heavily being that GM6 lacked complex abstract types like classes and structs. But it seems like having var pass-by-reference would be more efficient, I mean, the same reason that Java does it for objects.

Here's a little test for the ds_map_replace thing:
```gml
mymap = ds_map_create();
ds_map_replace(mymap, "test", 5);
show_message(string(ds_map_find_value(mymap, "test")));
```
GM5 on my pc shows "0"
ENIGMA shows empty dialog
GM:S shows "5"

So do you want to switch to the GM:S behavior? Are you sure nothing will break? It don't seem like anything could have been relying on the GM5 behavior.

Reply:
Taken care of, we'll get to ds_map_replaceanyway in a separate PR.
ab11647cea541924d11f236e2700c6e4c7714ac0
2effede47558be4b8b41a0e74e3e0ecac4050bd2